spacy_experimental/span_finder/span_finder_component.py

```diff
@@ -0,0 +1,323 @@
+        candidates_key: str = "span_finder_candidates",
+    ) -> None:
+        """Initialize the span boundary detector.
+        model (thinc.api.Model): The Thinc Model powering the pipeline component.
+        name (str): The component instance name, used to add entries to the
+            losses during training.
+        threshold (float): Minimum probability to consider a prediction
+            positive.
+        scorer (Optional[Callable]): The scoring method.
+        candidates_key (str): Name of the SpanGroup the span candidates are saved to
+        max_length (int): Max length of the produced spans (unlimited when set to 0)
+        min_length (int): Min length of the produced spans (unlimited when set to 0)
+        """
+        self.vocab = nlp
+        self.cfg = {
+            "threshold": threshold,
+            "max_length": max_length,
+            "min_length": min_length,
+            "candidates_key": candidates_key,
+        }
+        self.model = model
+        self.name = name
+        self.scorer = scorer
+
+    def predict(self, docs: Iterable[Doc]):
+        """Apply the pipeline's model to a batch of docs, without modifying them.
+        docs (Iterable[Doc]): The documents to predict.
+        RETURNS: The models prediction for each document.
+        """
+        scores = self.model.predict(docs)
+        return scores
+
+    def set_annotations(self, docs: Iterable[Doc], scores: Floats2d) -> None:
+        """Modify a batch of Doc objects, using pre-computed scores.
+        docs (Iterable[Doc]): The documents to modify.
+        scores: The scores to set, produced by SpanFinder predict method.
+        """
+        lengths = [len(doc) for doc in docs]
+
+        offset = 0
+        scores_per_doc = []
+        for length in lengths:
+            scores_per_doc.append(scores[offset : offset + length])
+            offset += length
+
+        for doc, doc_scores in zip(docs, scores_per_doc):
+            doc.spans[self.cfg["candidates_key"]] = []
+            starts = []
+            ends = []
+
+            for token, token_score in zip(doc, doc_scores):
+                if token_score[0] >= self.cfg["threshold"]:
+                    starts.append(token.i)
+                if token_score[1] >= self.cfg["threshold"]:
+                    ends.append(token.i)
+
+            for start in starts:
+                for end in ends:
+                    if start < end + 1:
+                        if (
+                            self.cfg["min_length"] <= 0
+                            or start >= self.cfg["min_length"]
+                        ):
+                            if (
+                                self.cfg["max_length"] <= 0
+                                or end <= self.cfg["max_length"]
+                            ):
+                                doc.spans[self.cfg["candidates_key"]].append(
+                                    doc[start : end + 1]
+                                )
```

Comment:
If you have two simple nested `if`s you can just use one `if` with `and` instead.

Right now this is comparing the `start`/`end` indices rather than the span length.

Remember to break out of the loop as soon as you see a span length that's longer than `max_length`.

Reply:
You're absolutely right! I don't know why it's taking me so long for this 😃 Updated in [ad9736e](https://github.com/explosion/spacy-experimental/pull/10/commits/ad9736efd0c998ba98e22abe1a8a63d8c08c783d)